Mathlib/Analysis/PSeries.lean

```diff
@@ -17,132 +17,218 @@ converges if and only if so does `∑ k, 2 ^ k f (2 ^ k)`. We prove this test in
 `NNReal.summable_condensed_iff` and `summable_condensed_iff_of_nonneg`, then use it to prove
 `summable_one_div_rpow`. After this transformation, a `p`-series turns into a geometric series.
 
-## TODO
-
-It should be easy to generalize arguments to Schlömilch's generalization of the Cauchy condensation
-test once we need it.
-
 ## Tags
 
 p-series, Cauchy condensation test
 -/
 
 open Filter
 
-open BigOperators ENNReal NNReal Topology
+open BigOperators ENNReal NNReal Topology Finset
 
 /-!
-### Cauchy condensation test
+### Schlömilch's generalization of the Cauchy condensation test
 
-In this section we prove the Cauchy condensation test: for `f : ℕ → ℝ≥0` or `f : ℕ → ℝ`,
-`∑ k, f k` converges if and only if so does `∑ k, 2 ^ k f (2 ^ k)`. Instead of giving a monolithic
-proof, we split it into a series of lemmas with explicit estimates of partial sums of each series in
-terms of the partial sums of the other series.
+In this section we prove the Schlömilch's generalization of the Cauchy condensation test:
+for `u : ℕ → ℕ` and `f : ℕ → ℝ≥0` or `f : ℕ → ℝ`, `∑ k, f k` converges if and only if
+so does `∑ k, (u (k + 1) - u k) * f (u k)`. Instead of giving a monolithic proof, we split it
+into a series of lemmas with explicit estimates of partial sums of each series in terms of the
+partial sums of the other series.
 -/
 
+/--
+A sequence `u` has the property that its ratio of successive differences is bounded
+when there is a positive real number `C` such that, for all n ∈ ℕ,
+(u (n + 2) - u (n + 1)) ≤ C * (u (n + 1) - u n)
+-/
+def succ_diff_bounded (u : ℕ → ℕ) : Prop :=
+  ∃ (C : ℝ≥0∞) (_ : C ≠ ∞) (_ : C > 0), ∀ (n : ℕ), (u (n + 2) - u (n + 1)) ≤ C * (u (n + 1) - u n)
```

Comment:
No need to use `C : ℝ≥0∞` if you are just going to assume `C ≠ ∞`, instead just use `ℝ≥0`. Also, with this notation notice that we can combine the existence of `C` with the assumption that it is positive.
```suggestion
  ∃ C > (0 : ℝ≥0), ∀ (n : ℕ), (u (n + 2) - u (n + 1)) ≤ C * (u (n + 1) - u n)
```
However, I think ultimately I would just pass `C` as an argument to `succ_diff_bounded` like we do with [LipschitzOnWith](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Topology/EMetricSpace/Lipschitz.html#LipschitzOnWith) instead of hiding it behind the existential. That way we can access the `C`.

Finally, I suspect that a lot of the pain you are dealing with is a result of coercions, some of which are caused by this definition. If you choose `C : ℕ` instead, it may make your life significantly easier.

Reply:
Thanks for the detailed review! I chose to use `∃ (C : ℝ≥0∞) (_ : C ≠ ∞) (_ : C > 0)` because I am getting some coercions errors that I do not really know how to deal with. I tried changing the definition to
```
def succ_diff_bounded (C : ℕ+) (u : ℕ → ℕ) : Prop :=
  ∀ (n : ℕ), (u (n + 2) - u (n + 1)) ≤ C * (u (n + 1) - u n)
```
which gives me a coercion error
```
application type mismatch
  mul_le_mul_of_nonneg_right (h_succ_diff k)
argument
  h_succ_diff k
has type
  u (k + 2) - u (k + 1) ≤ ↑C * (u (k + 1) - u k) : Prop
but is expected to have type
  ↑(u (k + 2)) - ↑(u (k + 1)) ≤ ↑↑C * (↑(u (k + 1)) - ↑(u k)) : Prop

```
 that I am not sure where to start to resolve.

A more fundamental issue that I have doubts about is that I defined f as `{f : ℕ → ℝ≥0∞}` in the namespace `Finset` instead of the original
```
variable {M : Type*} [OrderedAddCommMonoid M] {f : ℕ → M}
```
Also, if I defined it as `{f : ℕ → ℝ≥0}`, there will be some type mismatch error when doing
```
  refine' iSup_le fun n => (Finset.le_sum_schlomilch hf h_pos hu n).trans (add_le_add_left _ _)
```
in `ENNReal.le_tsum_schlomilch`.

Reply:
Minor remark: with the current naming convention, you should use `CamelCase` name for this definition.